### ISSUE TITLE
[BUG/ISSUE] Error in tropchem run directory generation 

### ISSUE THREAD

**Issue by ltmurray** [geoschem/geos-chem#617]
When using the new run directory generator in 13.0.0 with the tropchem mechanism, replace_colon_sep_val fails to correctly replace "--> STATE_PSC" with "false," and instead replaces it with "tfalse".

```
# ----- RESTART FIELDS ----------------------------------------
    --> GC_RESTART             :       true
    --> STATE_PSC              :       tfalse
    --> HEMCO_RESTART          :       true
```

**Comment by ltmurray** [comment]
This appears to only be a problem on macOS, not Linux.  I'll try to figure out why.

**Comment by yantosca** [comment]
I can also take a look.

**Comment by yantosca** [comment]
Thanks for bringing this up @ltmurray.  I wrote a test script to replicate the text replace:

```bash
#!/bin/bash

#=============================================================================
#### Define function to replace values in config files
#=============================================================================
replace_colon_sep_val() {
    KEY=$1
    VALUE=$2
    FILE=$3
    #printf '%-30s : %-20s %-20s\n' "${KEY//\\}" "${VALUE}" "${FILE}"

    # replace value in line starting with 'whitespace + key + whitespace + : +
    # whitespace + value' where whitespace is variable length including none

    # Doesn't work
    #sed -i -e "s|^\([\t ]*${KEY}[\t ]*:[\t ]*\).*|\1${VALUE}|" ${FILE}

    # Works
    sed -i -e "s|/^\([\t ]*${KEY}[\t ]*:[\t ]*\).*|\1${VALUE}|" ${FILE}
}

# Set up file that looks like the HEMCO_Config file
echo "--> GC_RESTART    : true" > file.txt
echo "--> STATE_PSC     : true" >> file.txt
echo "--> HEMCO_RESTART : true" >> file.txt

# Print file before text replace
echo "Before"
cat file.txt

# Replace the text
replace_colon_sep_val "--> STATE_PSC" false file.txt   

# Print file after text replace
echo
echo "After"
cat file.txt

exit 0
```
I found that the text `s|^\` at the beginning of the sed command needs to be changed to `s|/^\` (i.e. adding a slash in front of the ^.  That seems to work on both MacOS and Linux.  I'll verify that this works on both and then push a fix shortly.

**Comment by yantosca** [comment]
Actually I might need to tinker a bit more.  Stay tuned.

**Comment by yantosca** [comment]
So apparently MacOS (based off of BSD) sed [acts differently](https://unix.stackexchange.com/questions/13711/differences-between-sed-on-mac-osx-and-other-standard-sed) than Linux sed.  Also see: https://unix.stackexchange.com/questions/401905/bsd-sed-vs-gnu-sed-and-i

**Comment by yantosca** [comment]
Also see https://riptutorial.com/sed/topic/9436/bsd-macos-sed-vs--gnu-sed-vs--the-posix-sed-specification

**Comment by ltmurray** [comment]
Yeah, it also creates extra files relative to GNU sed (a backup of the original appended with -e).  The quick fix I'm using at the moment — since I'm not very good with RegEx — is the following at the end of createRunDir.sh

```
#-----------------------------------------------------------------
# Cleanup
#-----------------------------------------------------------------
# Quick fix for sed failure on macOS
sed -i -e 's/tfalse/false/g' ${rundir}/HEMCO_Config.rc
# sed leaves extra unnecessary files on macOS 
find ${rundir} -name "*-e" -exec rm {} \;
```

**Comment by yantosca** [comment]
This version works on both MacOS (BSD) and Linux:
```bash
#=============================================================================
#### Define function to replace values in config files
#=============================================================================
replace_colon_sep_val() {
    KEY=$1
    VALUE=$2
    FILE=$3
    #printf '%-30s : %-20s %-20s\n' "${KEY//\\}" "${VALUE}" "${FILE}"

    # replace value in line starting with 'whitespace + key + whitespace + : +
    # whitespace + value' where whitespace is variable length including none
    # NOTE: This sed command is valid on both MacOS and Linux (bmy, 2/23/21)
    TAB=$(printf "\t")
    sed -i'' -e "s|^\([${TAB} ]*${KEY}[${TAB} ]*:[${TAB} ]*\).*|\1${VALUE}|" ${FILE}

    # MacOS may leave a backup file with the -e extension, so remove it
    if [[ -f ${FILE}-e ]]; then
	rm ${FILE}-e
    fi
}
```
This uses the POSIX standard which is recognized by both MacOS/BSD and Linux.  Long story short, in MacOS, you cannot use any escaped characters (e.g. \t) in regular expressions.  The cheap solution is to print a tab character to a variable and then use that in the sed command where you would have used the \t.

Also, using the -e argument to Sed will append -e to the file name.  I have an IF statement to remove that.

Will push this shortly to dev/13.0.0-final.  Let me know if you run into other run-directory creation issues. There are a few sed commands used in these scripts so we might have to make a workaround.

**Comment by yantosca** [comment]
Also I wonder if it makes sense to make a new local sed function, something like
``` bash
sed_ie () {
    CMD=${1}
    FILE=${2}
    sed -i'' -e ${CMD} ${FILE}

    # MacOS may leave a backup file with the -e extension, so remove it
    if [[ -f ${FILE}-e ]]; then
	rm ${FILE}-e
    fi
}
```
as we may end up running into this issue elsewhere.

**Comment by yantosca** [comment]
FYI, the above script works, but you have to put quotes around the arguments, i.e.
```
sed -i'' -e "${CMD}" "${FILE}"
```
that will force the entire string to be passed to the sed command.

I am testing a fix with a set of integration tests and if they pass, I'll merge into dev/13.0.0-final for inclusion in 13.0.0.

**Comment by yantosca** [comment]
This is now fixed with these commits: 74b2d790278fa8672293ce26d7e9ad2df4eb6cc3, https://github.com/geoschem/GCClassic/commit/1ab1b989de38767aefccbd83632779e337fdc6a9, and https://github.com/geoschem/GCHP/commit/0eaa68be5d329d259fd18c30cf096987aef90421.  The GEOS-Chem "Classic" run directory generation script and integration test generation script should now work on both MacOS/Darwin and GNU/Linux.  This will ship with the 13.0.0 final release.
